crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,212 @@
+use rustpython_parser::ast::{Constant, Expr, ExprConstant, ExprSlice, ExprSubscript, Ranged};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use std::fmt;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for indexed access to lists, strings, tuples, bytes, and comprehensions
+/// using a type other than an integer or slice.
+///
+/// ## Why is this bad?
+/// Only integers or slices can be used as indices to these types. Using
+/// other types will result in a `TypeError` at runtime and a `SyntaxWarning` at
+/// import time.
+///
+/// ## Example
+/// ```python
+/// var = [1, 2, 3]["x"]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// var = [1, 2, 3][0]
+/// ```
+#[violation]
+pub struct InvalidIndexType {
+    value_type: String,
+    index_type: String,
+    is_slice: bool,
+}
+
+impl Violation for InvalidIndexType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidIndexType {
+            value_type,
+            index_type,
+            ..
+        } = self;
+        if self.is_slice {
```

Comment:
Nit: destructure `is_slice` above? Since you're already destructuring `value_type` and `index_type`. You may then have to do `if *is_slice`.

Reply:
Honestly I don't understand what difference the destructing makes at all, why do we do it here?

Reply:
The nice thing about the destructuring is that we can inline the variables directly in the `format!` call. This is valid:

```rust
let MultiValueRepeatedKeyLiteral { name } = self;
format!("Dictionary key literal `{name}` repeated")
```

However, this is not:

```rust
format!("Dictionary key literal `{self.name}` repeated")
```

Instead, you'd have to do:

```rust
format!("Dictionary key literal `{}` repeated", self.name)
```

That's the main benefit of destructuring in these violation implementations -- that we can inline values in the `format!` calls.

And then, if we're going to destructure _some_ members, IMO it's preferable to be consistent and destructure everything rather than mixing and matching.